File: website/parser/lexical.py
import re

class LexicalParser():

    sentence = ""
    length = 0
    text = ""
    
    def __init__(self, sentence):
        self.sentence = sentence
        
    def match(self, regex):
        m = re.match(r"^"+regex, self.sentence.lower())
        if(m == None):
            return None
        self.length = len(m.group(0))
        self.text = m.group(0)
        self.shift(self.length)
        return m.group(0)

    def shift(self, length):
        self.sentence = self.sentence[self.length:]

    def get_text(self):
        return self.text
    
    def get_lexeme(self):
        raise NotImplementedError

class SQLLexicalParser(LexicalParser):

    _query = 0
    _opening = 1
    _closing = 2
    _star = 3
    _dot = 4
    _comma = 5
    _quote = 6
    _double_quote = 7

    _equal = 8
    _not_equal = 9
    _less_e = 10
    _greater_e = 11
    _less = 12
    _greater = 13

    _select_distinct = 14
    _select = 15
    _from = 16
    _as = 17
    _where = 18
    _group_by = 19
    _having = 20
    _order_by = 21
    _asc = 22
    _desc = 23
    _and = 24
    _or = 25
    _not = 26
    _like = 27
    _in = 28
    _name = 29
    _number = 30
    
    def get_lexeme(self):
        if self.match("\(") : return self._opening
        if self.match("\)") : return self._closing
        if self.match("\*") : return self._star
        if self.match("\.") : return self._dot
        if self.match(",") : return self._comma
        if self.match("'") : return self._quote
        if self.match("\"") : return self._double_quote

        if self.match("=") : return self._equal
        if self.match("<>|!=") : return self._not_equal
        if self.match("<=") : return self._less_e
        if self.match(">=") : return self._greater_e
        if self.match("<") : return self._less
        if self.match(">") : return self._greater

        if self.match("select\s+distinct") : return self._select_distinct
        if self.match("select") : return self._select
        if self.match("from") : return self._from
        if self.match("as") : return self._as
        if self.match("where") : return self._where
        if self.match("group\s+by") : return self._group_by
        if self.match("having") : return self._having
        if self.match("order\s+by") : return self._order_by
        if self.match("asc") : return self._asc
        if self.match("desc") : return self._desc
        if self.match("and") : return self._and
        if self.match("or") : return self._or
        if self.match("not") : return self._not
        if self.match("like") : return self._like
        if self.match("in") : return self._in

        if self.match("[a-zA-Z0-9]+") : return self._name
        if self.match("[0-9]+") : return self._number
        if self.match("\s+") : return self.get_lexeme()

File: static/website/js/lexical.js
function LexicalParser(sentence) {
    this.sentence = sentence;
    this.length = 0;
    this.text = "";

    this.match = function(regex) {
	var m = this.sentence.toLowerCase().match(new RegExp('^'+regex));
	if(m == null) {
	    return null;
	} else {
	    m = m[0];
	}
	console.log("sentence: "+this.sentence);
	console.log("m: "+m);
	this.length = m.length;
	this.text = m;
	this.shift(this.length);
	return this.text;
    }

    this.shift = function(length) {
	this.sentence = this.sentence.substring(length);
    }

    this.get_text = function() {
	return this.text;
    }
}

function SQLLexicalParser(sentence) {
    LexicalParser.call(this, sentence);

    this._query = 0;
    this._opening = 1;
    this._closing = 2;
    this._star = 3;
    this._dot = 4;
    this._comma = 5;
    this._quote = 6;
    this._double_quote = 7;

    this._equal = 8;
    this._not_equal = 9;
    this._less_e = 10;
    this._greater_e = 11;
    this._less = 12;
    this._greater = 13;

    this._select_distinct = 14;
    this._select = 15;
    this._from = 16;
    this._as = 17;
    this._where = 18;
    this._group_by = 19;
    this._having = 20;
    this._order_by = 21;
    this._asc = 22;
    this._desc = 23;
    this._and = 24;
    this._or = 25;
    this._not = 26;
    this._like = 27;
    this._in = 28;
    this._name = 29;
    this._number = 30;

    this.color = ["", "red", "red", "blue", "black", "black", "blue", "blue",
		  "black", "black", "black", "black", "black", "black",
		  "green", "green", "green", "green", "green", "green", "green", "green", "green", "green", "green", "green", "green", "green", "green", "blue", "blue"];
    
    this.get_lexeme = function() {
	if (this.match("\\(")) { return this._opening; }
	if (this.match("\\)")) { return this._closing; }
        if (this.match("\\*")) { return this._star; }
        if (this.match("\\.")) { return this._dot; }
        if (this.match(",")) { return this._comma; }
        if (this.match("'")) { return this._quote; }
        if (this.match('\\"')) { return this._double_quote; }

        if (this.match("=")) { return this._equal; }
        if (this.match("<>|!=")) { return this._not_equal; }
        if (this.match("<=")) { return this._less_e; }
        if (this.match(">=")) { return this._greater_e; }
        if (this.match("<")) { return this._less; }
        if (this.match(">")) { return this._greater; }

        if (this.match("select\\s+distinct")) { return this._select_distinct; }
        if (this.match("select")) { return this._select; }
        if (this.match("from")) { return this._from; }
        if (this.match("as")) { return this._as; }
        if (this.match("where")) { return this._where; }
        if (this.match("group\\s+by")) { return this._group_by; }
        if (this.match("having")) { return this._having; }
        if (this.match("order\\s+by")) { return this._order_by; }
        if (this.match("asc")) { return this._asc; }
        if (this.match("desc")) { return this._desc; }
        if (this.match("and")) { return this._and; }
        if (this.match("or")) { return this._or; }
        if (this.match("not")) { return this._not; }
        if (this.match("like")) { return this._like; }
        if (this.match("in")) { return this._in; }

        if (this.match("[a-zA-Z0-9]+")) { return this._name; }
        if (this.match("[0-9]+")) { return this._number; }
        if (this.match("\\s+")) { return this.get_lexeme(); }
	return -1;
    }
}
